pkg/localprovider/localprovider.go

```diff
@@ -0,0 +1,103 @@
+package localprovider
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	upgradev1alpha1 "github.com/openshift/managed-upgrade-operator/pkg/apis/upgrade/v1alpha1"
+	"github.com/openshift/managed-upgrade-operator/util"
+)
+
+const UPGRADECONFIG_CR_NAME = "managed-upgrade-config"
+
+var log = logf.Log.WithName("upgradeconfig-localprovider")
+
+func New(c client.Client, name string) (*localProvider, error) {
+	return &localProvider{
+		client:  c,
+		cfgname: name,
+	}, nil
+}
+
+type localProvider struct {
+	client  client.Client
+	cfgname string
+}
+
+// Get checks the upgrade config on the cluster with matched name which is going to perform the upgrade
+func (l *localProvider) Get() ([]upgradev1alpha1.UpgradeConfigSpec, error) {
+	log.Info("Read the upgrade config from the cluster directly")
+
+	// Get the current UpgradeConfigs on the cluster
+	instances, err := fetchUpgradeConfigs(l.client)
+	if err != nil {
+		return nil, err
+	}
+
+	// Remove the upgradeConfig is it has been consumed
+	for _, u := range instances.Items {
```

Comment:
Instead of having the provider do the delete, the `upgradeconfigmanager` can take care of this for you.

If the UpgradeConfig returned from the fetch is for a completed upgrade, just return empty slice/nil. That'll cause the upgradeconfigmanager, which call the provider, to clean up the upgradeconfig.